components/layout/inline.rs

```diff
@@ -608,9 +634,21 @@ impl LineBreaker {
         let split_result = match fragment.calculate_split_position(available_inline_size,
                                                                    self.pending_line_is_empty()) {
             None => {
-                debug!("LineBreaker: fragment was unsplittable; deferring to next line");
-                self.work_list.push_front(fragment);
-                self.flush_current_line();
+                // We failed to split. Defer to the next line if we're allowed to; otherwise,
+                // rewind to the last line breaking opportunity.
+                if fragment_is_line_break_opportunity {
+                    debug!("LineBreaker: fragment was unsplittable; deferring to next line");
+                    self.work_list.push_front(fragment);
+                    self.flush_current_line();
+                } else if !self.split_line_at_last_known_good_position() {
```

Comment:
Personally I find this much easier to read if the `!` is removed from this condition, and the `if` and `else` cases swapped.  Up to you.